### DISCUSSION TITLE
direct ip connection not possible without relay server

### DISCUSSION THREAD

**Discussion by Phoenixium** [rustdesk/rustdesk#5054] [General]
### Bug Description

When trying a direct connection to devices on my local network with deactivated relay server i cant establish a session since version 1.2X
The Error message is "connection error: Closed manually by the peer", this appears after a split second after i accept the connection on the target client. I get a picture of the screen of the target machine in the background of the error message.

With version v1.9 i had no issues in connecting via IP without relay. In version 2.0/2.1 i would have to activate the relay service on the receiving side to connect via ip without error.

Connecting with v2.0(or v2.1) to a client that runs v1.9 also works fine without relay.
[source_rd.txt](https://github.com/rustdesk/rustdesk/files/12092729/source_rd.txt)
[target_rd.txt](https://github.com/rustdesk/rustdesk/files/12092730/target_rd.txt)

### How to Reproduce

Windows 10 with v2.1 Client with direct IP mode activated and relay service not started. 
stop-service = 'Y' and direct-server = 'Y' as config in rustdesk2.toml

Connecting to the target client via direct ip. (You will see a short screenshot of the target machine in the background of the error message). "connection error: Closed manually by the peer"

### Expected Behavior

Direct Connection working fine between clients like in/with version 1.1.9
No relay server or connection to the outside of the LAN needed. 

### Operating system(s) on local side and remote side

Windows 10 -> Windows 10

### RustDesk Version(s) on local side and remote side

1.2.1 -> 1.2.1

### Screenshots

![error](https://github.com/rustdesk/rustdesk/assets/139851827/2faffd0e-2621-4c6a-b488-8d217b7fadd9)

Error message + background of the target machine

**Comment by henry-Hon** [comment]
the same issue
https://github.com/rustdesk/rustdesk/issues/5062

**Comment by rustdesk** [comment]
What does "deactivated relay server" mean?

**Comment by rustdesk** [comment]
So you applied something undocumented.
` stop-service = 'Y' and direct-server = 'Y' as config in rustdesk2.toml`

Have to apploligize, we are very busy now, can not dive into this undocumented thing for the time being.

**Reply by henry-Hon** [reply]
I thought the config content is generated automatically by rustdesk process.
conf file path: 
C:\Users\xxx\AppData\Roaming\RustDesk\config\
file RustDesk2.toml will be modified when change has been made from UI
mine is as follow:

rendezvous_server = 'myself-hosted.com:21116'
nat_type = 1
serial = 0

[options]
relay-server = 'myself-hosted.com'
direct-server = 'Y'
stop-service = 'Y'
enable-rdp = 'N'
key = 'xxxxxx'
custom-rendezvous-server = 'myself-hosted.com'
local-ip-addr = '192.168.x.x'

So the following content is generated by rustdesk client process but not manually
stop-service = 'Y' and direct-server = 'Y'

**Reply by rustdesk** [reply]
Next time, another guy will report a new bug, `why the service is stopped, but direct connection is still ok.`

Which bug should we fix? should we fix the bug of 1.1.9, or the bug of 1.2?

**Reply by henry-Hon** [reply]
This discussion is talking about v1.2.x.
But you are right, to be clear, rustdesk-1.2.1-x86_64 and OS is win10

**Reply by rustdesk** [reply]
I will put a quick fix later, but the fix may be not what you wanna see, the direct connection will be completely disabled when `stop-service=Y`. https://github.com/rustdesk/rustdesk/commit/2ec773e9bf062cef84e8bec9e6af7825d2cae34e

**Reply by henry-Hon** [reply]
Thanks very much, I guess I understand what happened. When the service is disabled from UI, even the process still listen on 21118 (default port of direct IP mode), but the target does not work as expecting (as service is disabled).
I remember there exists a former request for seperately enabling direct-ip connection without connecting to relay-server, sorry for the accurate link. I can't remember well.
Most concern for disabling service may be that for unnecessary network connection (LAN scenario), unreachable case, or secure reason. 
That's fine, I have tested the work-around by changing the id-srv and relay-srv to a local ip instead, and enable service. In this way direct-ip connection is available as expected.
This issue is fixed in an efficient and acceptable way.
Thanks again.

**Reply by Phoenixium** [reply]
i didnt edit the config manually, what i simply did was activating direct-ip mode and deactivate the running relay-service (both in rustdesk itself), which is exactly the setup i had in version 1.1.9.

While its a way to deactivate the config combination to avoid further confusion and errors wouldnt it be better (if possible) to fix this to enable back this functionally and work case we had in version 1.1.9?  Its actually more or less this feature that made me really intressted in the software since its perfect for small home setups with headless servers and a work from home notebook.

**Reply by henry-Hon** [reply]
currently, my work-around is 
on target 
id server : 0.0.0.1
relay server: 0.0.0.1
key: whatever
apply config

then enable service and check updated RustDesk2.toml 

In this way client could connect target by direct ip

**Reply by haldi4803** [reply]
Thanks for that info.
Tried 127.0.0.1 without any API key and the Software kept Restarting^^

btw, 
IMHO Direct IP access should work even when Service is Offline.

**Reply by denny466** [reply]
Thanks henry-Hon! I thought i was the only one who rather uses VPN tunneling than Relays.

And thanks rustdesk for this great app! I understand that Relays is much, much simpler to setup and use but i believe that adding direct IP-only will exponentially increase awesomeness. Best regards.
